src/lib_ml_remla/preprocess.py

```diff
@@ -0,0 +1,70 @@
+"""
+Provides functions to preprocess data.
+
+"""
+import numpy as np
+from sklearn.preprocessing import LabelEncoder
+from keras._tf_keras.keras.preprocessing.text import Tokenizer
+from keras._tf_keras.keras.preprocessing.sequence import pad_sequences
+
+
+def split_data(train: list[str], test: list[str], val: list[str]) -> tuple[list[str], list[str], list[str], list[str], list[str], list[str]]:
+    """
+    Split the data into training, validation, and test sets.
+
+    Args:
+        train: List of strings containing the training data.
+        test: List of strings containing the test data.
+        val: List of strings containing the validation data.
+
+    Returns:
+        Tuple of the split data in the form (X_train, y_train, X_val, y_val, X_test, y_test).
+    """
+    raw_X_train = [line.split("\t")[1] for line in train]
+    raw_y_train = [line.split("\t")[0] for line in train]
+
+    raw_X_test = [line.split("\t")[1] for line in test]
+    raw_y_test = [line.split("\t")[0] for line in test]
+    
+    raw_X_val = [line.split("\t")[1] for line in val]
+    raw_y_val = [line.split("\t")[0] for line in val]
+
+    return raw_X_train, raw_y_train, raw_X_val, raw_y_val, raw_X_test, raw_y_test
+
+
+def preprocess_data(raw_X_train: list[str], raw_y_train: list[str],
+                    raw_X_val: list[str], raw_y_val: list[str],
+                    raw_X_test: list[str], raw_y_test: list[str], sequence_length: int = 200
+                    ) -> tuple[np.ndarray, np.ndarray, np.ndarray,
+                               np.ndarray, np.ndarray, np.ndarray, dict[str, int]]:
+    """
+    Preprocess the data for training the model.
+
+    Args:
+        raw_X_train: List of strings containing the training data.
+        raw_y_train: List of strings containing the training labels.
+        raw_X_val: List of strings containing the validation data.
+        raw_y_val: List of strings containing the validation labels.
+        raw_X_test: List of strings containing the test data.
+        raw_y_test: List of strings containing the test labels.
+        sequence_length: The length of the sequences to pad the data to.
+
+    Returns:
+        Tuple of preprocessed data in the form (X_train, y_train, X_val, y_val, X_test, y_test, char_index).
+
+    """
+    # Tokenize the dataset
+    tokenizer = Tokenizer(lower=True, char_level=True, oov_token='-n-')
+    tokenizer.fit_on_texts(raw_X_train + raw_X_val + raw_X_test)
```

Comment:
Fitting the tokenizer on train, val and test data might be overfitting.

Reply:
But in that case the logic changes a little bit. Is that ok for the assignment? Can we just use what we have or is that just not possible at all?

Reply:
I think its possible but it would mean the server would currently not use any functions provided by ml-lib, as the server cannot call any of these functions to preprocess the input. We probably need to do some refactoring later

Reply:
I will add this functionality in a different branch